### PR TITLE
Adjust interface in 'dofs' folder to new datatype 'types::fe_index'.

### PR DESCRIPTION
**Pull request by marcfehling**
Part of #13595.

I continued to adjust the interface to the new `types::fe_index` type and started from the `dofs` folder.

The only **_hard incompatibility_** I could find was with the usage of `DoFAccessor::get_active_fe_indices()`, as its return type changes to `std::set<types::fe_index>`. This function is mainly used internally, and I don't think it will clash with user code.

How consistent do we want to be with the use of `types::fe_index`? There are lots and lots of parts throughout the library that use `active_fe_indices` one way or another. Shall we keep the `unsigned int` interface and only store `active_fe_indices` internally with `types::fe_index`, or shall we progress towards a coherent datatype?

Right now I plan to continue to adjust the interface in other folders, like `hp`. For this I left some TODOs in the code introduced with this PR.

### CODE REVIEW

**Comment by drwells** [comment]
Do you know of any user projects which are both freely available and use hp-FEM? If we can check that no-one is using these features then I think its fine to have some minor compatibility breaks.

**Comment by marcfehling** [comment]
> Do you know of any user projects which are both freely available and use hp-FEM? If we can check that no-one is using these features then I think its fine to have some minor compatibility breaks.

I know that Agros is using the deal.II hp features, but the last time I checked they are using the hp interface older than deal.II version 9.0.

The only hard incompatibility I could find is just with one single function that is part of the public interface, but it is really only used internally.

I'll make sure to deprecate functions wherever possible, and to document all incompatibilities. Yes, they are a hassle to deal with, but I would still favor a more robust hp-codebase with stronger typing.

There will be more patches that follow this one.

File: source/dofs/dof_handler.cc [hunk 1]
```diff
@@ -1636,23 +1636,24 @@ namespace internal
...
           const std::set<unsigned int> children_fe_indices_set(
             children_fe_indices.begin(), children_fe_indices.end());
 
-          const unsigned int dominated_fe_index =
+          const types::fe_index dominated_fe_index =
             fe_collection.find_dominated_fe_extended(children_fe_indices_set,
                                                      /*codim=*/0);
 
-          Assert(dominated_fe_index != numbers::invalid_unsigned_int,
+          Assert(dominated_fe_index != numbers::invalid_fe_index,
```
Comment (bangerth) [inline source/dofs/dof_handler.cc:1656]: Nice catch!

File: source/dofs/dof_handler.cc [hunk 2]
```diff
@@ -2653,15 +2666,17 @@ void
 DoFHandler<dim, spacedim>::get_active_fe_indices(
   std::vector<unsigned int> &active_fe_indices) const
 {
-  active_fe_indices = get_active_fe_indices();
+  std::vector<types::fe_index> indices = get_active_fe_indices();
```
Comment (bangerth) [inline source/dofs/dof_handler.cc:2669]: make this `const`

**Review by bangerth — COMMENTED** [review]
Just a few minor comments.

File: include/deal.II/dofs/dof_handler.h [hunk 3]
```diff
@@ -580,45 +580,85 @@ class DoFHandler : public Subscriptor
   operator=(const DoFHandler &) = delete;
 
   /**
-   * Go through the triangulation and set the active FE indices of all
-   * locally owned cells to the values given in @p active_fe_indices.
+   * For each locally owned cell, set the active finite element index to the
+   * corresponding value given in @p active_fe_indices. Indices must be in the
+   * order how we iterate over active cells.
```
[the remark this answers is not in the document]
Comment (marcfehling) [reply include/deal.II/dofs/dof_handler.h:585]: `
Vector entries corresponding to active cells that are not locally owned are ignored.
`

I covered this part in the next paragraph.

`
Active FE indices will only be set for locally owned cells. Ghost and artificial cells will be ignored; no active FE index will be assigned to them.
`

File: include/deal.II/dofs/dof_handler.h [hunk 4]
```diff
@@ -580,45 +580,85 @@ class DoFHandler : public Subscriptor
...
-   * Go through the triangulation and set the future FE indices of all
-   * locally owned cells to the values given in @p future_fe_indices.
-   * Cells corresponding to numbers::invalid_fe_index will be skipped.
+   * For each locally owned cell, set the future finite element index to the
+   * corresponding value given in @p future_fe_indices. Indices must be in the
+   * order how we iterate over active cells.
+   *
+   * Future FE indices will only be set for locally owned cells. Ghost and
+   * artificial cells will be ignored; no future FE index will be assigned to
+   * them.
```
[the remark this answers is not in the document]
Comment (marcfehling) [reply include/deal.II/dofs/dof_handler.h:645]: See below
`
The vector @p future_fe_indices is assumed to have as many entries as there are active cells.
`

File: include/deal.II/dofs/dof_handler.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (bangerth) [reply include/deal.II/dofs/dof_handler.h]: That's not quite the same to what I wanted to say. You're saying you're not touching these *cells*. I'm saying that there must still be entries in the vector for them anyway.

File: source/dofs/dof_handler.cc [hunk 5]
```diff
@@ -2891,16 +2906,16 @@ DoFHandler<dim, spacedim>::pre_distributed_transfer_action()
 
   active_fe_index_transfer->cell_data_transfer = std::make_unique<
     parallel::distributed::
-      CellDataTransfer<dim, spacedim, std::vector<unsigned int>>>(
+      CellDataTransfer<dim, spacedim, std::vector<types::fe_index>>>(
```
Comment (marcfehling) [inline source/dofs/dof_handler.cc:2909]: This means we also need to update the version in the metafile for parallel distributed serialization. I haven't done this just yet, as I wanted to open more PRs after this one.

But we can update the version already in this PR. What do you prefer?

**Comment by marcfehling** [comment]
_ping_

**Review by drwells — APPROVED** [review]
I think the small break is fine. If there aren't further comments I'll merge in a day or so.

File: include/deal.II/dofs/dof_handler.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (marcfehling) [reply include/deal.II/dofs/dof_handler.h]: Okay. I get your point. I'll add these lines.

**Comment by marcfehling** [comment]
I've updated the documentation text according to @bangerth's comment.

@drwells I interpreted your 'thumbs up' that you'd prefer to bump the serialization version, so I did it in this PR in a separate commit. We need to do it anyways, so why not now when the change goes in effect.

**Comment by peterrum** [comment]
Why does this PR influence restart files in the non-hp context? See https://github.com/lethe-cfd/lethe/pull/623.

FYI @blaisb

**Comment by marcfehling** [comment]
This should not influence non-hp applications. We had to bump the serialization version as we do not have a way to distinguish between hp and non-hp on the serialized data level. Sadly, this means that compatibility to serialized data from 9.4 is lost.

@blaisb -- It should've been fine if you just increased the version number from 4 to 5 in https://github.com/lethe-cfd/lethe/pull/623 for non-hp applications. The other changes in your PR are unrelated to this patch.

**Pull request merged**